### PR TITLE
feat: AMO Manager V2 with unified debt system

### PR DESCRIPTION
**Pull request by Git-on-my-level**
## Summary

Implements a comprehensive AMO (Algorithmic Market Operations) debt system with unified accounting for both stable and collateral AMO operations. This PR introduces the AMO Manager V2 architecture that provides atomic operations with value conservation invariants.

### 🏗️ New Contracts

**AmoDebtToken** - Transfer-restricted ERC20 receipt token
- Dynamic naming: "dTRINITY AMO Receipt" with symbols "amo-dUSD" or "amo-dS"
- Transfer restrictions via allowlist to prevent manipulation
- Role-based minting/burning for secure debt management

**AmoManagerV2** - Unified AMO operations manager
- **Stable AMO**: Atomic mint/burn of dUSD with matching debt tokens
- **Collateral AMO**: Borrow/repay with vault value conservation invariants
- Configurable tolerance for multi-decimal asset support
- Admin-controlled multisig and allowlists

### ⚡ Technical Highlights

- **Atomic Operations**: All AMO operations are atomic with automatic rollback on failure
- **Value Conservation**: Collateral operations enforce `vault.totalValue()` remains unchanged (within tolerance)
- **Multi-decimal Support**: Handles assets with 6/8/18 decimals correctly
- **Security**: ReentrancyGuard protection, role-based access control, allowlist restrictions
- **Gas Optimized**: Uses EnumerableSet for efficient allowlist operations, immutable references

### 📦 Deployment Infrastructure

- Complete deployment scripts with idempotent operations
- Automatic configuration of roles, allowlists, and oracle prices
- Governance migration scripts for transitioning control to multisig
- Comprehensive test fixtures that validate deployment configuration

## Testing

- ✅ Comprehensive unit tests for both contracts
- ✅ Integration tests validating deployment scripts
- ✅ Edge case handling and security validation
- ✅ Multi-decimal asset testing (USDC 6 decimals, standard 18 decimals)

### Contracts
- `contracts/dstable/AmoDebtToken.sol` - Receipt token implementation
- `contracts/dstable/AmoManagerV2.sol` - Unified AMO manager

### Deployment
- `deploy/18_amo_manager_v2_and_debt/01_deploy_amo_manager_v2_and_debt.ts`
- `deploy/18_amo_manager_v2_and_debt/02_configure_amo_system.ts`
- `deploy/18_amo_manager_v2_and_debt/03_migrate_roles_to_governance.ts`

### Tests
- `test/dstable/AmoManagerV2.spec.ts` - Unit tests
- `test/dstable/AmoManagerV2.deployment.spec.ts` - Deployment validation tests
- `test/dstable/fixtures.ts` - Updated with AMO V2 fixtures

### Configuration
- `typescript/deploy-ids.ts` - Added AMO V2 deployment IDs
- `hardhat.config.override.js` - Compiler configuration for AMO contracts

## Checklist

- [x] Code follows project conventions and style
- [x] All tests pass
- [x] Linting passes without errors
- [x] Documentation updated
- [x] Deployment scripts tested
- [x] Security considerations addressed

## Notes

The system is designed to be upgradeable and extensible. Future enhancements could include:
- Additional AMO strategies
- Enhanced reporting and analytics
- Integration with other protocol components

🤖 Generated with Claude Code

### CODE REVIEW

**Comment by Git-on-my-level** [comment]
## ✅ AMO Debt System Alignment Complete

This PR has been updated with the requested changes to align the AMO Debt System with the one-sided invariant design.

### Latest Changes (commit 42782e3)

#### 🔄 Role Renames Implemented
- `AMO_MINTER_ROLE` → `AMO_DECREASE_ROLE` 
- `AMO_BORROWER_ROLE` → `AMO_INCREASE_ROLE`

#### 🛡️ Security Enhancements Added
- `DebtTokenProhibited()` guard in `borrowTo()` and `repayFrom()`
- Prevents debt token from being used as collateral

#### 📊 One-Sided Invariant Confirmed
- Tests updated to validate: `postValue + tolerance >= preValue`
- Allows vault gains while preventing losses beyond tolerance

#### ✅ Comprehensive Testing
- Added slippage protection tests
- Added adversarial debt token tests  
- **All 116 tests passing**

#### 🚨 Critical Fix Applied
- Fixed role name mismatch in dS system migration script
- Deployment scripts now consistent across both dUSD and dS systems

### QA Results

Comprehensive QA performed by specialized agents:
- **Security Review**: ✅ No vulnerabilities found
- **Functional Review**: ✅ Math and logic verified correct
- **Test Coverage**: ✅ All critical paths covered
- **Deployment Review**: ✅ Scripts validated and fixed

This PR is ready for final review and merge.

**Comment by Git-on-my-level** [comment]
Initial audit pass
image

File: contracts/dstable/AmoManagerV2.sol [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+     * @param amount The amount of dUSD to mint (debt tokens minted will equal this in base value)
+     * @param wallet The AMO wallet to receive minted dUSD
+     * @dev Only callable by AMO_INCREASE_ROLE. Ensures debt minted equals dUSD value
+     */
+    function increaseAmoSupply(uint256 amount, address wallet) external onlyRole(AMO_INCREASE_ROLE) nonReentrant {
+        // Validate inputs
+        if (!_allowedAmoWallets.contains(wallet)) {
+            revert UnsupportedAmoWallet(wallet);
+        }
+
```
Comment (Lange1337) [inline contracts/dstable/AmoManagerV2.sol:123]:
```suggestion
_assertDebtPeg();
```

File: contracts/dstable/AmoManagerV2.sol [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+     * @param amount The amount of dUSD to burn (debt tokens burned will equal this in base value)
+     * @param wallet The AMO wallet to pull dUSD from (must approve manager)
+     * @dev Only callable by AMO_DECREASE_ROLE. Ensures debt burned equals dUSD value
+     */
+    function decreaseAmoSupply(uint256 amount, address wallet) external onlyRole(AMO_DECREASE_ROLE) nonReentrant {
+        // Validate inputs
+        if (!_allowedAmoWallets.contains(wallet)) {
+            revert UnsupportedAmoWallet(wallet);
+        }
+
```
Comment (Lange1337) [inline contracts/dstable/AmoManagerV2.sol:167]:
```suggestion
_assertDebtPeg();
```

File: contracts/dstable/AmoManagerV2.sol [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+        if (!_allowedAmoWallets.contains(wallet)) {
+            revert UnsupportedAmoWallet(wallet);
+        }
+        if (!CollateralVault(collateralVault).isCollateralSupported(asset)) {
+            revert UnsupportedCollateral(asset);
+        }
+        if (asset == address(debtToken)) {
+            revert DebtTokenProhibited();
+        }
+
```
Comment (Lange1337) [inline contracts/dstable/AmoManagerV2.sol:229]:
```suggestion
_assertDebtPeg();
```

File: contracts/dstable/AmoManagerV2.sol [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external onlyRole(AMO_DECREASE_ROLE) nonReentrant {
+        IERC20Permit(asset).permit(wallet, address(this), amount, deadline, v, r, s);
+        repayFrom(wallet, asset, amount, maxDebtBurned);
+    }
+
+    /* Helper Functions */
+
```
Comment (Lange1337) [inline contracts/dstable/AmoManagerV2.sol:331]:
```suggestion
  /**
     * @notice Asserts that the debt token oracle is correctly pegged at 1.0
     * @dev Prevents operations when debt token price deviates from baseCurrencyUnit
     */
    function _assertDebtPeg() internal view {
        uint256 debtTokenPrice = oracle.getAssetPrice(address(debtToken));
        if (debtTokenPrice != baseCurrencyUnit) {
            revert InvariantViolation(baseCurrencyUnit, debtTokenPrice);
        }
    }

```

**Review by Lange1337 — CHANGES_REQUESTED** [review]
We should add a depeg check, even though we have it hardcoded in the Oracle. 
We have multiple places where we need to make changes if we modify decimals. Let's say we make Fraxtal follow the same decimals as the rest of the chains. This way makes sure if we missed one, it won't break, and since it's only when the multisig chooses to do an action, the gas cost is worth it imho
